sig-api-machinery/design-docs/admission-control-webhooks.md

```diff
@@ -0,0 +1,934 @@
+1.  We have experience with Webhooks API already to give confidence in the API.  The registration API will be quite similar except in the responses.
+1.  There is a strong community demand for something that satisfies a mutating case.
+
+
+## Plan for Existing Initializer-Clients 
+
+After the release of 1.9, we will advise users who currently use initializers to:
+
+
+
+*   Move to Webhooks if their use case fits that model well.
+*   Provide SIG-API-Machinery with feedback if Initializers is a better fit. 
+
+We will continue to support Initializers as an Alpha API in 1.9.
+
+We will point out https://github.com/caesarxuchao/example-webhook-admission-controller and extend it to support mutating webhooks, and as that Initializer docs point to it (e.g. https://github.com/kelseyhightower/kubernetes-initializer-tutorial)
+
+We will clearly document the reasons for each and how users should decide which to use.
+
+
+## Monitorability 
+
+There should be prometheus variables to show:
+
+
+
+*   API operation latency
+    *   Overall
+    *   By webhook name
+*   API response codes
+    *   Overall
+    *   By webhook name.
+
+Adding a webhook dynamically adds a key to a map-valued prometheus metric. Webhook host process authors should consider how to make their webhook host monitorable: while eventually we hope to offer a set of best practices around this, for the initial release we won't have requirements here.
+
+
+## API Changes 
+
+GenericAdmissionWebhook Admission Controller is split and renamed.
+
+
+
+*   One is called `MutatingAdmissionWebhook`
+*   The other is called `ValidatingAdmissionWebhook`
+*   Splitting them allows them to appear in different places in the `--admission-control` flag's order.
+
+ExternalAdmissionHookConfiguration API is split and renamed.
+
+
+
+*   One is called  `MutatingAdmissionWebhookConfiguration`
+*   The other is called `ValidatingAdmissionWebhookConfiguration`
+*   Splitting them:
+    *   makes it clear what the order is when some items don't have both flavors,
+    *   enforces mutate-before-validate,
+    *   better allows declarative update of the config than one big list with an implied partition point
+
+The `ValidatingAdmissionWebhookConfiguration` stays the same as `ExternalAdmissionHookConfiguration` except it moves to v1beta1.
+
+The `MutatingAdmissionWebhookConfiguration` is the same API as `ValidatingAdmissionWebhookConfiguration`.  It is only visible via the v1beta1 version.
+
+The Group/Version called 
+
+`admissionregistration.k8s.io/v1alpha1` with kinds
+
+InitializerConfiguration and ExternalAdmissionHookConfiguration.
+
+InitializerConfiguration will not join `admissionregistration.k8s.io/v1beta1` at this time.
+
+Any webhooks that register with v1alpha1 may or may not be surprised when they start getting Versioned data.  But we don't make any promises for Alpha, and detecting which API was used to register is too complex.
```

Comment:
I think this should be stated more strongly: Sending internal data was a bug that must be corrected for transition to beta.  We acknowledge that it will break current alpha consumers, but the bug must be corrected for proper consumption.

Reply:
added 30% more strength